Implement `SVOFilter` type, make statistics generic

cgarling opened the pull request:
Following #21, which implemented the ability to query the SVO filter service, this PR implements an `SVOFilter` type that contains the transmission curve and the metadata returned by SVO. It is implemented as an encapsulated `PhotometricFilter` instance and a dictionary (2 total fields). A few dictionary methods are passed through to the `SVOFilter` type so you can do things like 

```
julia> filt = get_filter("2MASS/2MASS.J", :Vega);
julia> filt["ZeroPoint"] # Can retrieve metadata directly
```
to retrieve metadata from the contained dictionary, but I had trouble finding the full dictionary API documented anywhere so I could be missing things. I tried to add the stuff I most commonly use (`getindex`, `haskey`, `keys`, `values`).

To better support this new type, I made many of the base methods of the package operate on `AbstractFilter` where before they required a concrete `PhotometricFilter`. This required two new accessor methods, `name`, and `detector_type` to retrieve the relevant information from subtypes of `AbstractFilter`. 

A few unrelated changes:
 - I made `Base.getindex(f::AbstractFilter, i::Int) = (wave(f)[i], throughput(f)[i])` rather than just returning the throughput as it was doing previously; not sure in what circumstance you would *only* want to iterate over the throughput rather than the wavelength and throughput together.
 - The wavelength field of `PhotometricFilter` is now a Unitful vector, rather than plain numbers, to ensure that `wave(::AbstractFilter)` does not need to make a copy -- otherwise the above method would make a copy on every access. Previously, calling `wave` would allocate a copy of the wavelength field because it had to add units to it.

Happy to have feedback on this! I tried to add as few additional methods as possible, but now that we have a filter type with metadata, maybe we should add a few more methods (e.g., `zeropoint(::AbstractFilter)`) to provide consistent access to this information (when it is available -- probably return `missing` when unavailable). Tagging @lucasvalenzuela since they were interested in #21.

cgarling commented:
Couldn't sleep so added a few more methods ...
`integrate` computes basically $\overline{f_\lambda}$ as defined, for example in [pyphot's docs](https://mfouesneau.github.io/pyphot/photometry.html) and many other places. Not sure about the method name, not attached to `integrate` at all. I can verify the values against the [SVO blackbody synthetic photometry](https://svo2.cab.inta-csic.es/svo/theory/newov2/syph.php?model=bbody) for both photon counters and energy counters -- using the [Planck.jl](https://github.com/JuliaAstro/Planck.jl?tab=readme-ov-file) package for blackbody radiation we have for a photon counting system (SDSS u filter)

```julia
using Planck: blackbody
using PhotometricFilters: wave, integrate, SDSS_u
using Unitful
filt = SDSS_u()
Flam = blackbody.(u"erg/s/cm^2/angstrom", wave(filt), 5000u"K")
integrate(filt, wave(filt), Flam)
```
```
648628.4038417398 erg Å^-1 cm^-2 s^-1
```

which can be verified against [SVO's synthetic photometry](https://svo2.cab.inta-csic.es/svo/theory/newov2/phot.php?model=bbody&fid=99&filter=SLOAN/SDSS.u&format=ascii) which gives 648596.9309048

For an energy counting system (2MASS J filter)

```julia
using Planck: blackbody
using PhotometricFilters: wave, integrate, get_filter
using Unitful
filt = get_filter("2MASS/2MASS.J")
Flam = blackbody.(u"erg/s/cm^2/angstrom", wave(filt), 5000u"K")
integrate(filt, wave(filt), Flam)
```
```
445507.8175161778 erg Å^-1 cm^-2 s^-1
```

which can be compared against [SVO's synthetic photometry](https://svo2.cab.inta-csic.es/svo/theory/newov2/phot.php?model=bbody&fid=99&filter=2MASS/2MASS.J&format=ascii) which gives 446898.39843973

`F_nu` and `F_lambda` are supposed to convert between spectral flux density and spectral energy density units. I'm basing this off the [SVO API paper](https://www.ivoa.net/documents/Notes/SVOFPS/NOTE-SVOFPS-1.0.20121015.pdf) which writes
![image](https://github.com/user-attachments/assets/260e4697-d704-43d1-95e7-106c88c2a631)
which seems mostly the same as the "Jansky definition" given in [pyphot's docs](https://mfouesneau.github.io/pyphot/photometry.html#jansky-definition) as well,
![image](https://github.com/user-attachments/assets/bed34c84-0727-4496-9466-4812aa96c162)
though pyphot uses the pivot wavelength and SVO uses the effective wavelength. It's been a **long** time since I had to think about these spectral units so please let me know if I made a mistake with these.

lucasvalenzuela commented:
The PR looks great! I haven't had time to actually test the functionality yet, but I spent some time reading through the diffs and left one comment on the speed of light value.

Some further questions/comments:

- I just noticed `AbstractFilter{T}` is a subtype of `AbstractArray{T}`. What is the background for this? I would then expect `filt[i]` to return an element of type `T`, but that is not the case now. Is there any disadvantage to removing this subtyping?
- Does it make sense to implement the wavelength methods for `SVOFilter` to return the metadata directly? The advantage is direct access to the values, the disadvantage is that it can differ by a little compared to the actual computation implemented here, and the difference can be very large when the value is directly specified for the filter in the metadata (I'll reference the 2MASS.J filter here again as an example: https://svo2.cab.inta-csic.es/theory/fps/index.php?id=2MASS/2MASS.J&&mode=browse&gname=2MASS&gname2=2MASS#filter).
- For the computation of magnitudes and converting back and forth between `F_nu` and `F_lambda`, it looks like SVO uses the "reference wavelength", which they by default take as the pivot wavelength, but which I assume is replaced with the specified reference wavelength when available. We could perhaps add a `reference_wave` function to pull the specified value from the metadata for `SVOFilter` and otherwise calls `pivot_wave` for `AbstractFilter`. I think that would then allow for a consistent conversion for magnitudes etc.
- I think a zeropoint function sounds like a good idea, maybe as `zeropoint(filt, system::Symbol)` with `system` being one of `:Vega`, `:AB`, `:ST`.

cgarling commented on src/core.jl:
```diff
@@ -311,19 +312,149 @@ julia> apply(f, λ_u, flux) == f.(λ_u)
+
+where ``c`` is the speed of light in *m/s* and ``\\lambda_p`` is the pivot wavelength (`λpivot`) in *Angstroms*. If providing an [`AbstractFilter`](@ref PhotometricFilters.AbstractFilter) as the second argument, the pivot wavelength will be automatically computed with [`pivot_wavelength`](@ref PhotometricFilters.pivot_wavelength).
+"""
+F_nu(Fλ::SpectralFluxDensity, λpivot) = inv(2.9979246) * 1e5 * _ustrip(u"angstrom", λpivot)^2 * ustrip(u"erg/s/cm^2/angstrom", Fλ) * u"Jy"
```
For numeric constants I usually prefer to use `Rational`s as they have nice promotion properties  -- e.g., if the arguments to F_nu would be Float32, it would currently promote to Float64, but replacing the prefactor with a `Rational` would return a Float32. The Rational equivalent of the leading constant is `rationalize(1e5 / 2.99792458) == 25370985150//760603`. I would like to switch the implementation to use 25370985150//760603  -- I'll add a note of what this "magic constant" is for future reference.

lucasvalenzuela commented:
Good idea on using rational numbers! Only the comment then has to be without the decimal point: `Prefactor is (299792458 / 10^8) / 1e5`      (instead of 2.99...)

On the `AbstractVector` question: My feeling is that it is not intuitive that a "photometric filter" would be a "vector" in the mathematical sense. Even without subtyping `AbstractVector` it is possible to implement `getindex` and friends as well as the iteration methods to have the `for (lambda, through) in filt` functionality. If we do stick to `AbstractVector{T}`, then per the documentation `T` should be the element type: "Supertype for one-dimensional arrays (or array-like types) with elements of type T" (https://docs.julialang.org/en/v1/base/arrays/#Base.AbstractVector), so something along the lines of `Tuple{eltype(wave), eltype(through)}`. I suspect there are no big practical differences on the user side between subtyping it or not in the end, though.

lucasvalenzuela commented:
Here is the PR for removing the AbstractVector subtyping and adding back the functionality for looping etc.: https://github.com/cgarling/PhotometricFilters.jl/pull/5

cgarling commented:
@lucasvalenzuela I renamed the methods you suggested and added new methods to calculate zeropoints in Vega, AB, and ST mags. There are doctests for the new functions and generally I have compared against pyphot (e.g., the HST WFC3 F110W tests I compared against [this example notebook](https://github.com/mfouesneau/pyphot/blob/master/examples/Sun_Vega.ipynb)) and the SVO zero points. Probably for SVOFilters we should have specific methods that query the saved metadata for these quantities, but I wanted to get the generic methods in first.

lucasvalenzuela commented:
This looks great! It's really nice to have such an easy way of computing a magnitude from an SED and a filter!

Here are all the thoughts I had while going through the code and doing some tests with the new functionality (sorry for another long list like this...):

- Name changes: I think the changes to the previously introduced names are nice, and I do like the `mean_flux_density` even more now.
- New names: At first I was surprised that the new functions all start with the magnitude system, so for example `AB_zeropoint` instead of `zeropoint_AB`, do you have thoughts on which one would be better? I was thinking I would intuitively first type the "thing" I want like "mag..." and then autocomplete would show me I have the three magnitude system options for it.
- Similar to how `get_filter` takes a second argument for the mag system, I think it would be handy to have the same for `zeropoint_Jy`, `zeropoint_flux`, `zeropoint_mag`, and `magnitude` (at least I think that would be better than only `mag`, since I would see myself wanting to name my variables `mag = magnitude(filt, :AB)` for example). If we do that than maybe only export those and keep the specific variations public but unexported?
- Playing around with the zeropoints, I noticed that the Vega zeropoint is different from that of the SVO website (of course not by much, but still quite significant compared to AB and ST):

For example:
```julia
filt = get_filter("Euclid/VIS.vis", :AB)
Vega_zeropoint_flux(filt) # 1.6927306764889102e-9 erg Å^-1 cm^-2 s^-1
```
Whereas on the site we find (https://svo2.cab.inta-csic.es/theory/fps/index.php?id=Euclid/VIS.vis&&mode=browse&gname=Euclid&gname2=VIS#filter): 1.6846e-9 (erg/cm2/s/A)
Do you know where that comes from? Different Vega spectra standards I suppose?

- I was thinking it would be nice to have some kind of inverse function to go from a magnitude back to the corresponding flux in erg/cm^2/s/A or Jy as wished. On the naming side maybe something like `mag2flux_Jy(filt, mag, :AB)`?
- For `SVOFilter` objects, do you think it would be good to always by default take their proposed zeropoints? On the one hand it seems like a good idea to take their zeropoints if they are specified by the filter provider, on the other hand the SVO filter site has the AB Jy zeropoint fixed at 3631 Jy instead of the more correct definition (even though it is a small difference).
- When further implementing the `SVOFilter` part, I think it would be cool to also support the more exotic magnitude conversions, that is `Asinh` and `Linear` (in addition to the standard `Pogson`).

Unfortunately I am traveling quite a bit at the moment, but when that is done I'd be more than happy to contribute to all of this as well. In any case I am really excited about all of the new features that are going to be available in this package very soon!

cgarling commented:
You could allow the Vega standard to be configurable using the Symbol design as well. I think there are multiple ways you could do it, but one way is to use common buffers and a setter method, like

```julia
using Unitful, UnitfulAstro
using DataDeps
using PhotometricFilters # checked out to this PR

# Buffers in package scope
const vega_wavelength = Vector{typeof(1.0f0u"angstrom")}(undef, 0)
const vega_flux = Vector{typeof(1.0f0u"erg/cm^2/s/angstrom")}(undef, 0)

# `f` argument would be the standard you want to use, for now we use old file
function set_Vega(f::AbstractString = datadep"vega/vega.hd5") 
    # Flush buffers
    empty!(vega_wavelength)
    empty!(vega_flux)
    # Whatever code you need to read `f` here; for now just use PhotometricFilters.Vega()
    # wave, flux = read(f) 
    wave, flux = PhotometricFilters.Vega()
    for i in eachindex(wave, flux)
        push!(vega_wavelength, wave[i])
        push!(vega_flux, flux[i])
    end
end

# New Vega() method requires no computation, would just return buffers
Vega() = (vega_wavelength, vega_flux) 
```

I think this type of design is not thread-safe so generally I try to avoid it. But it is nice to configure "I want to use this Vega spectrum" once, and then have `Vega()` always use that model so you don't have to pass around a Vega object to every method. 

**On second thought**, passing around the Vega object probably isn't that onerous as you probably just do something like `vg = Vega("alpha_lyr_stis_003")` and then pass it around like `zeropoint_Jy(filt, vg)`. So the dispatch idea is growing on me.

I'm not sure how we mesh this with the zeropoint metadata in the SVOFilter type exactly -- my current preference would be to just always calculate our own (so not use the zeropoint metadata) but we could revisit this point in the future. Or maybe doing `zeropoint_Jy(f::SVOFilter, :Vega)` (with a Symbol second argument) returns the SVO metadata entry while `zeropoint_Jy(f::SVOFilter, vv::Vega)` with a Vega type as the second argument computes it from scratch.

cgarling commented:
`zeropoint...` functions renamed and dispatch-based zeropoint methods added. For now I'm just forwarding the calls so `zeropoint_flux(f::AbstractFilter, ::AB) = zeropoint_flux_AB(f)`. I added docstrings for the new methods but kept the old ones around as well as I wasn't sure how best to combine them. I figure once the implementation looks how we want it, we can more clearly decide what the docs should look like

lucasvalenzuela commented:
For some reason I thought I had already responded to the latest changes. All of it looks really good, I am quite excited about the current state! Looking back at this I think it is a good decision doing things with the Vega/AB/ST dispatch methods. On the code side, I think this is actually ready to go, I think everything else I have in mind would be better reserved for a follow-up PR. Do you have anything else you think would be good for this PR?

You also raised the point on the docs - do you know if it's possible to see them through Github for an individual PR, or do I have to compile it locally? Just from looking over them in the markdown files, it already looks fine to me.

cgarling commented:
I refactored the zeropoint and magnitude code to only use the dispatch-based methods, which allowed me to consolidate the docstrings nicely. I think this is now good to merge. If you have any comments @lucasvalenzuela let me know! 

In the event I don't hear from you, I think I will merge this over the weekend to try to keep things moving forward. I would like to work on #27 next but I don't want to start on that until we merge this. We can always revise things in future PRs if there are things you'd like to change!

lucasvalenzuela commented:
The changes look great and I think this PR is good to go! Good call on moving to the name `MagnitudeSystem`.

Thanks so much for putting in all the work here @cgarling!
